### ISSUE TITLE
Expression tree is too large (maximum depth 1000) with `including(all:)`

### ISSUE THREAD

**Issue by cstephens-sysco** [groue/GRDB.swift#871]
### What did you do?

- When loading 1000+ objects from a series of joined tables via RxGRDB observeAll we are seeing SQL errors and a failure to produce the objects:

```Expression tree is too large (maximum depth 1000) in "SELECT *, "id" AS "grdb_id", "opco" AS "grdb_opco" FROM "productPricingAppliedDiscount" WHERE ("customerId" = ?) AND ("opco" = ?) AND ((("id" = ?) AND ("opco" = ?)) OR (("id" = ?) AND ("opco" = ?)) OR (("id" = ?) AND ("opco" = ?)) OR (("id" = ?) AND ("opco" = ?)) OR (("id" = ?) AND ("opco" = ?)) OR (("id" = ?) AND ("opco" = ?)) OR (("id" = ?) AND ("opco" = ?)) OR (("id" = ?) AND ("opco" = ?)) OR (("id" = ?) AND ("opco" = ?)) OR (("id" = ?) AND ("opco" = ?)) OR (("id" = ?) AND ("opco" = ?)) OR (("id" = ?) AND ("opco" = ?)) OR (("id" = ?) AND ("opco" = ?)) OR (("id" = ?) AND ("opco" = ?)) OR (("id" = ?) AND ("opco" = ?)) OR (("id" = ?) AND ("opco" = ?))...```

### What did you expect to happen?

- The objects to be returned

### What happened instead?

- SQL errors and no objects are pulled from the tables

### Environment

**GRDB flavor(s):** (GRDB, SQLCipher, Custom SQLite build?)
GRDB
RxGRDB

**GRDB version:**
GRDB 4.9.0
RxGRDB 0.18.0

**Installation method:** (CocoaPods, SPM, manual?)
Cocoapods

**Xcode version:**
Xcode 12.1

**Swift version:**
5.3

**Platform(s) running GRDB:** (iOS, macOS, watchOS?)
iOS

**macOS version running Xcode:**
10.15.6

**Comment by groue** [comment]
Hello @cstephens-sysco,

It looks like you've reached the limits of the SQLite parser.

This may be solvable at the GRDB level, I don't know yet. Do you have any clue at what's happening? The requests you run, the generated SQL, maybe an alternative to this SQL that would avoid the error?

**Comment by groue** [comment]
@cstephens-sysco, actually it would be nice if you would provide a [Minimal, Reproducible Example](https://stackoverflow.com/help/minimal-reproducible-example):

```swift
struct ProductPricingAppliedDiscount { ... }

try DatabaseQueue().write { db in
    // Create tables
    // Insert some data
    // Build and run the failing request
}
```

**Comment by cstephens-sysco** [comment]
@groue I'm not quite sure what is happening just yet. I'm going to try and track down the SQL and see if I can do something alternative there. This didn't used to be a problem until we added two additional tables and added those to the original fetching.

Once I narrow it down I can try to find a minimal reproducible scenario for you. 

I did see that the maximum depth can be changed in SQLLite, is this something that is exposed in GRDB or possible?

**Comment by groue** [comment]
> Once I narrow it down I can try to find a minimal reproducible scenario for you.

Yes, please. I tried to reproduce but could not.

> I did see that the maximum depth can be changed in SQLLite, is this something that is exposed in GRDB or possible?

That's an interesting option, thanks for spotting it! But let's first see how to trigger the error.

I know that GRDB does currently not deal with SQLite limits. The max number of arguments is huge on the stock SQLite that ships with Apple systems, so I did never quite look into it. This issue had me learn about the depth of the expression tree.

This deep expression tree may be:
- the result of poor SQL generated by GRDB, that needs to be generated differently.
- the result of not-so-bad SQL generated by GRDB, but that has to be split in chunks.
- the result of a database schema that has GRDB generate inefficient SQL that can only be fixed at the schema level.
- maybe something we did not expect.

Without a reproducing case that is close to your actual database schema, it will be hard to say!

**Comment by groue** [comment]
> is this something that is exposed in GRDB or possible?

For the sake of completeness:

- https://www.sqlite.org/c3ref/c_limit_attached.html
- https://www.sqlite.org/c3ref/limit.html

This could give (not tested):

```swift
var config = Configuration()
config.prepareDatabase { db in
    sqlite3_limit(db.sqliteConnection, SQLITE_LIMIT_EXPR_DEPTH, 1000)
}
let dbQueue = DatabaseQueue(path, configuration: config)
```

But this will not prevent the error for good. And my own interest, as well as the interest of all GRDB users, is that we grab a better understanding of the topic ;-)

**Comment by cstephens-sysco** [comment]
Thanks, I agree, we still should track this down. From my understanding you might only be able to set the value lower during runtime, not higher. To set the value higher, do you have to compile it in the SQL library itself?

So I was able to to output the select statement, but what is odd to me is it is lacking all the `including(all: ...)` calls I have in there:
```SELECT "productListItem".*, "product".*, "productPricing".*, "productOrderable".* FROM "productListItem" JOIN "product" ON ("product"."id" = "productListItem"."supc") AND ("product"."opco" = "productListItem"."opco") JOIN "productPricing" ON ("productPricing"."id" = "product"."id") AND ("productPricing"."opco" = "product"."opco") AND ("productPricing"."customerId" = ?) AND ("productPricing"."opco" = ?) JOIN "productOrderable" ON ("productOrderable"."id" = "product"."id") AND ("productOrderable"."opco" = "product"."opco") AND ("productOrderable"."customerId" = ?) AND ("productOrderable"."opco" = ?) WHERE ("productListItem"."listId" = ?) AND ("productListItem"."customerId" = ?) AND ("productListItem"."opco" = ?) ORDER BY "productListItem"."orderedDate" DESC```

The request itself looks more like this:
```let sortColumn = listType == .purchaseHistory ?
                    ProductListItemColumn.orderedDate.reversed :
                    ProductListItemColumn.customSequence
                
                let request = ProductListItem.filter(
                    ProductListItemColumn.listId == listId &&
                    ProductListItemColumn.customerId == accountId &&
                    ProductListItemColumn.opco == opco
                )
                .order(sortColumn)
                .including(required: ProductListItem.product)
                .including(
                    required: ProductListItem.productPricing.filter(
                        ProductPricingColumn.customerId == accountId &&
                        ProductPricingColumn.opco == opco
                    )
                )
                .including(
                    all: ProductListItem.productPricingAppliedDiscounts.filter(
                        ProductPricingAppliedDiscountColumn.customerId == accountId &&
                        ProductPricingAppliedDiscountColumn.opco == opco
                    )
                )
                .including(
                    all: ProductListItem.productPricingTieredPrices.filter(
                        ProductPricingTieredPriceColumn.customerId == accountId &&
                        ProductPricingTieredPriceColumn.opco == opco
                    )
                    .forKey("productPricingTieredPricesWithDiscounts")
                    .including(
                        all: ProductPricingTieredPrice.productPricingTieredAppliedDiscounts.filter(
                            ProductPricingTieredAppliedDiscountsColumn.customerId == accountId &&
                            ProductPricingTieredAppliedDiscountsColumn.opco == opco
                        )
                    )
                )
                .including(
                    required: ProductListItem.productOrderable.filter(
                        ProductOrderableColumn.customerId == accountId &&
                        ProductOrderableColumn.opco == opco
                    )
                )
                .asRequest(of: ProductListItemProduct.self)```

**Comment by groue** [comment]
This still not quite help me reproduce the issue ;-) That's too much guessing. I know it's long to build a minimal and reproducible example, and I'd totally understand this looks like a charge for you, but this issue won't move forward without it.

**Comment by cstephens-sysco** [comment]
I understand, let me start creating a project to reproduce this for you.

In other news, it looks like this code 
```sqlite3_limit(db.sqliteConnection, SQLITE_LIMIT_EXPR_DEPTH, 1000)```
Can only be used to set the limit lower than 1000. You have to change the limit to something higher in the compilation of SQL itself.

**Comment by groue** [comment]
> In other news, it looks like this code
sqlite3_limit(db.sqliteConnection, SQLITE_LIMIT_EXPR_DEPTH, 1000)
Can only be used to set the limit lower than 1000. You have to change the limit to something higher in the compilation of SQL itself.

Good to know 👍

**Comment by cstephens-sysco** [comment]
Hi @groue , I was able to bring our schema over, trimming all the excess and get it to reproduce in a sample project for you. 

Everything lives in the ViewController file and you will notice the output after run is the same message. Just open the project file in Xcode and run.

I appreciate you taking a look and please let me know if you need something else from me.

[GRDBLimitTest.zip](https://github.com/groue/GRDB.swift/files/5549206/GRDBLimitTest.zip)

**Comment by groue** [comment]
Thank you @cstephens-sysco. I'll have a look. Your own opinion on https://github.com/groue/GRDB.swift/issues/871#issuecomment-728122025 will be welcomed, if you have one.

**Comment by cstephens-sysco** [comment]
It looks like the SQL generated by using associations `including(all: ...)` is too complex, maybe more complex than it should. I can't really see the full SQL it outputs when I try, the only parts I see in the prepared statement are the `including(required: ...)` so I'm not sure what it happening behind the scenes.

**Comment by groue** [comment]
### Diagnostic

The issue is triggered by eager loading of a has-many association based on a compound foreign key:

```swift
parent.including(all: children)
```

Eager loading of has-many associations has GRDB generally generate the following SQL:

```sql
SELECT * FROM parent
SELECT * FROM child WHERE parentID IN (?, ?, ...)
```

For compound foreign keys, this translates to:

```sql
SELECT * FROM parent
SELECT * FROM child WHERE ((a = ?) AND (b = ?)) OR ((a = ?) AND (b = ?)) OR ...
```

This `((a = ?) AND (b = ?)) OR ((a = ?) AND (b = ?)) OR ...` expression has SQLite fail with the `Expression tree is too large` error when it contains too many elements.

Minimal reproducing code, based on the sample project uploaded by @cstephens-sysco:

```swift
struct ProductListItem: Codable, FetchableRecord, PersistableRecord {
    let supc: Int
    let opco: String
    static let productPricingAppliedDiscounts = hasMany(ProductPricingAppliedDiscount.self)
}

struct ProductPricingAppliedDiscount: Codable, FetchableRecord, PersistableRecord {
    let id: Int
    let opco: String
}

try DatabaseQueue().write { db in
    
    // MARK: - Create Tables
    
    try db.create(table: "productListItem") { table in
        table.column("supc", .integer).notNull()
        table.column("opco", .integer).notNull()
        table.primaryKey(["supc", "opco"])
    }
    
    try db.create(table: "productPricingAppliedDiscount") { table in
        table.column("id", .text).notNull()
        table.column("opco", .text).notNull()
        table.foreignKey(["id", "opco"], references: "productListItem")
    }
    
    // MARK: - Insert Data
    
    let opco = "1"
    for index in 0..<2000 {
        try ProductListItem(supc: index, opco: opco).save(db)
        try ProductPricingAppliedDiscount(id: index, opco: opco).save(db)
    }
    
    // MARK: - Fetch Data
    
    let request = ProductListItem
        .including(all: ProductListItem.productPricingAppliedDiscounts)
        .asRequest(of: Row.self)
    
    db.trace { event in
        if case let .statement(statement) = event {
            print(statement.sql)
        }
    }
    let _ = try request.fetchAll(db)
}
```

**Comment by groue** [comment]
### Analysis

> The issue is triggered by eager loading of a has-many association based on a compound foreign key:
> 
> ```swift
> parent.including(all: children)
> ```
> 
> Eager loading of has-many associations has GRDB generally generate the following SQL:
> 
> ```sql
> SELECT * FROM parent
> SELECT * FROM child WHERE parentID IN (?, ?, ...)
> ```
> 
> For compound foreign keys, this translates to:
> 
> ```sql
> SELECT * FROM parent
> SELECT * FROM child WHERE ((a = ?) AND (b = ?)) OR ((a = ?) AND (b = ?)) OR ...
> ```

What can we say about this?

All children for all parents are loaded in a single request. GRDB then dispatches children to their parents in memory. This is the general strategy for avoiding the [N + 1 problem](https://stackoverflow.com/questions/97197).

Both children requests are *correct*.

The first one (`parentID IN (?, ?, ...)`) has benefits:

- It is not subject to the `Expression tree is too large` error.
- It is limited by the maximum number of arguments, [SQLITE_LIMIT_VARIABLE_NUMBER](https://www.sqlite.org/limits.html), which is fairly large (500000 on recent Apple systems).
- It can trigger an index on the parent table.

The second one (`((a = ?) AND (b = ?)) OR ((a = ?) AND (b = ?)) OR ...`) has defects:

- It can trigger the `Expression tree is too large` error.
- It can't trigger any index on the parent table.

In order to avoid the `Expression tree is too large` error, we can:

1. Split the request into several ones, each one below the [SQLITE_LIMIT_EXPR_DEPTH](https://www.sqlite.org/limits.html) limit:
    
    ```sql
    SELECT * FROM parent
    SELECT * FROM child WHERE ((a = ?) AND (b = ?)) OR ((a = ?) AND (b = ?))
    SELECT * FROM child WHERE ((a = ?) AND (b = ?)) OR ((a = ?) AND (b = ?))
    ...
    SELECT * FROM child WHERE (a = ?) AND (b = ?)
    ```

2. Use [row values](https://www.sqlite.org/rowvalue.html) (SQLite 3.15+ / iOS 10.3+ / macOS 10.13+):
    
    ```sql
    SELECT * FROM parent
    SELECT * FROM child WHERE (a, b) IN ((?, ?), (?, ?), ...)
    ```
    
    Row values may avoid the `Expression tree is too large` error (not tested yet).
    
    Row values may trigger an index on the parent table (not tested yet).
    
    Row values are not available on all systems supported by GRDB (iOS 10.0+ / macOS 10.10+), so the splitting strategy remains necessary.

3. Something else?

**Comment by groue** [comment]
> Use [row values](https://www.sqlite.org/rowvalue.html) (SQLite 3.15+ / iOS 10.3+ / macOS 10.13+):
>     
> ```sql
> SELECT * FROM parent
> SELECT * FROM child WHERE (a, b) IN ((?, ?), (?, ?), ...)
> ```

Darn! This SQL is not valid. According to [the doc](https://www.sqlite.org/rowvalue.html#row_value_comparisons):

> For a row-value `IN` operator, [...] the right-hand side [...] must be a subquery expression.
> 
> ```sql
> CREATE TABLE t2(x,y,z);
> INSERT INTO t2(x,y,z) VALUES(1,2,3),(2,3,4),(1,NULL,5);
> SELECT
>    (1,2,3) IN (SELECT * FROM t2),  -- 1
>    (7,8,9) IN (SELECT * FROM t2),  -- 0
>    (1,3,5) IN (SELECT * FROM t2);  -- NULL
> ```

So we're only left with the splitting technique. A technique which, as explained above, has defects, the worse being that it can not trigger indexes on the parent table.

In the current state of my knowledge, I have to conclude that *to-many associations based on a compound foreign key are inherently inefficient in `including(all:)` requests*.

---

Now, @cstephens-sysco, let's get practical, so that we both rationally spend our time.

1. Do you *expect* your app to meet this issue in production?
2. Is this issue a blocker, or did you find a workaround?

If this issue is a blocker, you have some options:

1. Wait for me or somebody else to implement the splitting technique.
2. Join the GRDB contributors, and implement the splitting technique yourself (I'll provide appropriate guidance).
3. Refactor your database schema so that all foreign keys to `(spuc, opco)` pairs are refactored as foreign keys to a single indexed column.
4. Do not use `including(all:)` in order to fetch the data needed by your app (you'll perform several other requests and merge the results in your application code).

I strongly suggest you seriously study the third option, because if your database can really get big, indexes can really help. *It is not true that one can ignore the intrinsic performance characteristics of an underlying database.*

Please tell me what are your next steps, I'll react accordingly.

**Comment by groue** [comment]
> ```sql
> SELECT * FROM child WHERE (a, b) IN ((?, ?), (?, ?), ...)
> ```
>
> Darn! This SQL is not valid.

But this one is:

```sql
SELECT * FROM child WHERE (a, b) = (?, ?) OR (a, b) = (?, ?) OR ...
```

Yet experiments show that such a rewrite does not solve the `Expression tree is too large` error :-( So previous conclusions and questions still hold.

**Comment by cstephens-sysco** [comment]
Hi @groue , thank you for your in-depth analysis. To answer your questions:

1. Unfortunately our app did have this problem in production
2. I did find a workaround, although I'm not sure the performance will be the best it could be. I had already implemented what you mentioned in Option 4. By fetching the rows separately using a `contains` on the `supc` and filter on the `opco` and `customerId`.

`try ProductPricingAppliedDiscount.filter(
        ProductPricingAppliedDiscountColumn.customerId == accountId &&
        ProductPricingAppliedDiscountColumn.opco == opco
)
.filter(productListItemSupcs.contains(ProductPricingAppliedDiscountColumn.id))
.fetchAll(db)`

The unfortunate part is having to do basically do 4 fetches to get all the information I need, so I'm not sure if there would be a faster or more performant way of doing this?

As for option 3, I'm not sure I'm following, do you mean instead of have multiple column primary keys to instead have a single primary key with the ids in a combined unique single key?

**Comment by groue** [comment]
Thanks, @cstephens-sysco

> I did find a workaround

Excellent 👍 

> although I'm not sure the performance will be the best it could be.

I'm not sure the GRDB change that would avoid your error (splitting the request) could do really better, because the two-columns foreign key *will* generate inefficient SQL (see the description on the N+1 problem and the only solution I know [above](#issuecomment-731738851)). Only benchmarks could tell.

> The unfortunate part is having to do basically do 4 fetches to get all the information I need

Well, the error you get is unfortunate. It will be fixed.

But remember that GRDB will *never* fulfill all applications needs. It is *normal* to perform several requests sometimes. All GRDB features used to be... not there yet, and thus implemented by the host applications. There will always be apps that want more 😉 Fortunately, database observation apis don't care about the number of requests you have to run 😅 

> As for option 3, I'm not sure I'm following, do you mean instead of have multiple column primary keys to instead have a single primary key with the ids in a combined unique single key?

I don't quite know how to combine them. I was thinking about an new table that has three columns: a autoincremented primary key (for efficient foreign keys), and a (spuc, opco) secondary unique key. But since you already have a workaround, I don't think you or I will investigate further such a schema refactoring.

All right. Now there only remains the "Expression tree is too large" error to solve in the case of `including(all:)` with compound foreign keys. This issue will remain open until it is.

**Comment by groue** [comment]
I've been looking at fetching children in batches in `parent.including(all: Parent.children)`, in order to simplify the generated SQL, and avoid the `Expression tree is too large` error.

Precisely speaking, I talk about turning:

```sql
SELECT * FROM child WHERE ((a = ?) AND (b = ?)) OR ((a = ?) AND (b = ?)) OR ...
```

into:

```sql
--                        <- limit the complexity of the WHERE clause ->
SELECT * FROM child WHERE ((a = ?) AND (b = ?)) OR ((a = ?) AND (b = ?))
SELECT * FROM child WHERE ((a = ?) AND (b = ?)) OR ((a = ?) AND (b = ?))
...
SELECT * FROM child WHERE (a = ?) AND (b = ?)
```

Instead of fetching children for all parents in one request, we fetch children for *batches of parents*, in several requests.

**There are several difficulties to deal with.**

```swift
let request = parent.including(all: Parent.children)
```

The above request could be automatically batched: the batch size is a direct function of `SQLITE_LIMIT_EXPR_DEPTH`. Yet:

**Difficulty**: Can we assume that SQLite inner algorithms that limit the complexity of expressions are stable? If not, GRDB automatic batch size may generate errors with future releases of SQLite, iOS, macOS, etc.

We also have to deal with:

```swift
let request = parent.including(all: Parent.children.filter(Column("foo") > 12 && ...))
let request = parent.including(all: Parent.children.filter(sql: "ARBITRARY SQL"))
```

I could test that children filtering takes its toll of the request complexity, and require that the batch size must be reduced.

**Difficulty**: How to compute the toll on expression complexity brought by those filters?

At this point, all hopes of being able to provide an automatic batch size were lost. Instead, I was considering letting the user choose the batch size (and assume the full responsibility of any error):

```swift
// Let the user deal with it ¯\_(ツ)_/¯
let request = parent.including(all: Parent.children.inBatches(of: 500))
```

Then came the death blow:

```swift
// 😱 ordering ☠️
let request = parent.including(all: Parent.children.inBatches(of: 500).order(...))
```

**Difficulty**: Since children are fetched in batches, we have to restore ordering afterwards. But we don't know how to reimplement in Swift the SQLite ordering.

---

At this point, I realize how **hard** this problem is 😅

We could use a [CTE](https://sqlite.org/lang_with.html) and [row values](https://www.sqlite.org/rowvalue.html). This would drastically reduce the SQL complexity:

```sql
-- Parent.filter(...).including(all: Parent.children.order(...))

SELECT * FROM parent WHERE ...

WITH parent AS (SELECT a, b FROM parent WHERE ...)
SELECT * FROM child
WHERE (a, b) IN (SELECT * FROM parent)
ORDER BY ...
```

But I'm not even sure this technique deals well with parents joined with many other associations in complex ways.

At this point, the amount of necessary work is much bigger than a plain enhancement. CTEs have been on my radar [for a while](https://forums.swift.org/t/using-ctes-with-clause/39327/9), and I know this is not a trivial topic, and it needs further maturing.

**Conclusion**

The documentation will be updated so that it makes clear that `including(all:)` does not support compound foreign keys when there are many source records. One should instead refactor the schema with single-column foreign keys, or prefetch manually (as you did).

This may look disappointing. But one should not underestimate the value of a *simple mental model*.

**Comment by groue** [comment]
v5.2.0 has shipped with updated documentation: https://github.com/groue/GRDB.swift/blob/master/CHANGELOG.md#520

Thanks @cstephens-sysco for reporting this caveat. Now let's wait for CTEs to mature!

**Comment by groue** [comment]
> v5.2.0 has shipped with updated documentation: https://github.com/groue/GRDB.swift/blob/master/CHANGELOG.md#520
> 
> Thanks @cstephens-sysco for reporting this caveat. Now let's wait for CTEs to mature!

The fix has been found: https://github.com/groue/GRDB.swift/pull/885

Next version will ship with it.

**Comment by groue** [comment]
The fix has shipped in v5.3.0

**Comment by cstephens-sysco** [comment]
> The fix has shipped in v5.3.0

Thanks again for all your work!
